daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1020,14 +1020,20 @@ private[lf] object SBuiltin {
               templateId,
               machine.committers,
               cbMissing = _ => machine.tryHandleException(),
-              cbPresent = { coinst =>
-                // Note that we cannot throw in this continuation -- instead
-                // set the control appropriately which will crash the machine
-                // correctly later.
-                if (coinst.template != templateId)
-                  machine.ctrl = SEWronglyTypeContractId(coid, templateId, coinst.template)
-                else
-                  machine.ctrl = SEImportValue(coinst.arg.value)
+              cbPresent = {
+                case V.ContractInst(actualTmplId, V.VersionedValue(version, arg), _) =>
+                  // Note that we cannot throw in this continuation -- instead
+                  // set the control appropriately which will crash the machine
+                  // correctly later.
+                  machine.ctrl =
+                    if (actualTmplId != templateId)
+                      SEDamlException(DamlEWronglyTypedContract(coid, templateId, actualTmplId))
+                    else if (!machine.inputValueVersions.contains(version))
```

Comment:
This will be run on every value we get from the ledger and should hence rather be fast. Do we have any idea how expensive `contains` is? Does it get more expensive as we add more versions?

Reply:
This is not particularly optimized. However I think this should be pretty negligible in comparison of the three passes of translations the value pass through (proto to java, java to LF Value, LF Value to Speedy values).

This gets also more expensive as we add more versions. If this becomes an issue,  we could just add a cache in `VersionRange#contains`

Reply:
I see. Thanks for the explanation.